README
NAME
    Business::DK::FI - validation of Danish FI numbers

VERSION
    The documentation describes version 0.02

SYNOPSIS
        use Business::DK::FI qw(validate validateFI generate);
    
        if (validate('026840149965328')) {
            print "026840149965328 is valid\n";
        }
    
        my $fi_number = generate(1);
    
        if ($fi_number eq '000000000000018') {
            print "we have a FI number\n";
        }
    
DESCRIPTION
    FI numbers are numbers used on GIRO payment forms. These can be used to
    do online payments in banks or at in physical banks or post offices.

    The module currently only supports FI numbers in the following series:

    * 71
    * 75

SUBROUTINES AND METHODS
  validate
    Takes a single argument. 15 digit FI number. Returns true (1) or false
    (0) indicating whether the provided parameter adheres to specification.

  validateFI
    Less intrusive exported variation of validate. It is actually validate
    which is wrapping validateFI.

  generate
    Simple FI generation method. Takes an arbitrary number adhering to the
    following requirements:

    * length between 1 and 14
    * value between 1 and 99999999999999

    Returns a valid FI number.

  PRIVATE SUBROUTINES AND METHODS
  _calculate_checksum
    This method calculates a checksum, it takes a single number as parameter
    and returns the calculated checksum.

  _calculate_sum
    This method calculates a sum it takes a number and a reference to an
    array of control cifers. It calculates a single sum based on the number
    and the control cifer and returns this.

DIAGNOSTICS
    All methods die if their API is not respected. Method calls can with
    success be wrapped in Try::Tiny or `eval' blocks.

CONFIGURATION AND ENVIRONMENT
    The module requires no special configuration or environment.

DEPENDENCIES
    * Params::Validate
    * Readonly
    * Exporter
    * English

BUGS AND LIMITATIONS
    This module has no known bugs or limitations.

BUG REPORTING
    Please report issues via CPAN RT:

    * http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-DK-FI

    or by sending mail to

    * `<bug-Business-DK-FI@rt.cpan.org>'

TEST AND QUALITY
  TEST COVERAGE
            ---------------------------- ------ ------ ------ ------ ------ ------ ------
            File                           stmt   bran   cond    sub    pod   time  total
            ---------------------------- ------ ------ ------ ------ ------ ------ ------
            blib/lib/Business/DK/FI.pm    100.0  100.0    n/a  100.0  100.0   34.8  100.0
            ...b/Class/Business/DK/FI.pm   97.6   83.3    n/a  100.0  100.0   65.2   96.9
            Total                          99.1   90.0    n/a  100.0  100.0  100.0   98.7
            ---------------------------- ------ ------ ------ ------ ------ ------ ------

QUALITY AND CODING STANDARD
    The code passes Perl::Critic tests at severity 1 (*brutal*) with a set
    of policies disabled. please see t/perlcriticrc and the list below:

    * Perl::Critic::Policy::ValuesAndExpressions::ProhibitConstantPragma,
    please see:
    http://logiclab.jira.com/wiki/display/OPEN/Perl-Critic-Policy-ValuesAndE
    xpressions-ProhibitConstantPragma
    * Perl::Critic::Policy::Documentation::RequirePodLinksIncludeText, this
    is listed in the TODO it requires numerous changes to the distribution
    POD.
    * Perl::Critic::Policy::NamingConventions::Capitalization, validateFI is
    exported both as validateFI and validate and FI is our used acronym so
    we try to stick to this
    * Perl::Critic::Policy::ControlStructures::ProhibitCStyleForLoops, this
    is for the main algorithm, it was easier to do with a C-style for loop
    * Perl::Critic::Policy::Subroutines::RequireArgUnpacking, this is due to
    the way: Params::Validate is used
    * Perl::Critic::Policy::ValuesAndExpressions::ProhibitMagicNumbers, we
    use long control numbers
    * Perl::Critic:.:Policy::Variables::ProhibitPunctuationVars

TODO
    Please see the distribution TODO file also and the distribution road map
    at:
    http://logiclab.jira.com/browse/BDKFI#selectedTab=com.atlassian.jira.plu
    gin.system.project%3Aroadmap-panel

SEE ALSO
    * http://www.pbs.dk/
    * Try::Tiny
    * Business::DK::CVR
    * Business::DK::CPR
    * Business::DK::PO
    * Business::DK::Postalcode
    * Business::DK::Phonenumber

AUTHOR
    Jonas B. Nielsen, (jonasbn) - `<jonasbn@cpan.org>'

COPYRIGHT
    Business-DK-FI and related is (C) by Jonas B. Nielsen, (jonasbn)
    2009-2011

LICENSE
    Business-DK-FI and related is released under the artistic license

    The distribution is licensed under the Artistic License, as specified by
    the Artistic file in the standard perl distribution
    (http://dev.perl.org/licenses/artistic.html).



Changes
Changes file for Business::DK::FI

$Id: Changes 8172 2013-07-31 16:45:55Z jonasbn $

Project change log:
http://logiclab.jia.com/browse/BDKFI#selectedTab=com.atlassian.jira.plugin.system.project%3Achangelog-panel

0.06 2013-07-31

- Bug fix release, update only necessary if you are using an older Perl
  demonstrating this particular problem

- Qualifying use of ReadOnly more explicitly in an attempt to address report
  on failing test described in BDKFI-22

  https://logiclab.jira.com/browse/BDKFI-22


0.05 2013-07-29

- Maintenance release, update not necessary

- Fixed up Changes file as part of my Questhub quest adhering to the standard
  described in: CPAN::Changes::Spec [BDKFI-21]

  Quest: http://questhub.io/realm/perl/quest/51f0fd8e18ba7dcf4b000040
  Spec: https://metacpan.org/module/CPAN::Changes::Spec
  
  https://logiclab.jira.com/browse/BDKFI-21
  

0.04 2011-12-03

- Bug fix release, update recommended for users of 32-bit perl

- Bug fix release re-addressing previous attempts at fixing bug (see 0.03). The issue seemed to be one of 
  architecture and where I have only developed on 64-bit enabled platforms using 64-bit enabled perl I could
  not replicate the issue, which was related to Perl's sprintf on 32-bit perl. I now use string indications (s)
  instead of signed integer (d) and it seems to work.
  http://logicLAB.jira.com/browse/BDKFI-20


0.03 2011-11-23

- Bug fix release, update only necessary if you are using an older Perl
  demonstrating this particular problem

- Attempt at addressing failing tests, reported by CPAN testers. I have not been
  able to replicate the errors locally, but I have implemented a new strategy for
  addressing the same problem [BDK-17]
  http://logicLAB.jira.com/browse/BDKFI-17
  

0.02 2011-04-30

- Maintenance release, update not necessary

- Addressed some issues with Perl::Critic tests, introducing dependency on Try::Tiny

- Addressed issue with t/critic.t test being run even without Test::Perl::Critic being 
  installed, updated to contemporary boilerplate [BDKFI-16]
  http://logicLAB.jira.com/browse/BDKFI-16


0.01 2011-04-22

- Initial release


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4005
use ExtUtils::MakeMaker;
WriteMakefile
(
  'NAME' => 'Business::DK::FI',
  'VERSION_FROM' => 'lib/Business/DK/FI.pm',
  'PREREQ_PM' => {
                   'Carp' => '0',
                   'Class::InsideOut' => '0',
                   'Exporter' => '0',
                   'Params::Validate' => '0',
                   'Readonly' => '0',
                   'Test::Exception' => 0,
                   'Try::Tiny' => '0'
                 },
  'INSTALLDIRS' => 'site',
  'EXE_FILES' => [],
  'PL_FILES' => {},
  'test' => {
              'TESTS' => 't/*.t t/.svn/*.t t/.svn/prop-base/*.t t/.svn/props/*.t t/.svn/text-base/*.t t/.svn/tmp/*.t t/.svn/tmp/prop-base/*.t t/.svn/tmp/props/*.t t/.svn/tmp/text-base/*.t t/Business/*.t t/Business/.svn/*.t t/Business/.svn/prop-base/*.t t/Business/.svn/props/*.t t/Business/.svn/text-base/*.t t/Business/.svn/tmp/*.t t/Business/.svn/tmp/prop-base/*.t t/Business/.svn/tmp/props/*.t t/Business/.svn/tmp/text-base/*.t t/Business/DK/*.t t/Business/DK/.svn/*.t t/Business/DK/.svn/prop-base/*.t t/Business/DK/.svn/props/*.t t/Business/DK/.svn/text-base/*.t t/Business/DK/.svn/tmp/*.t t/Business/DK/.svn/tmp/prop-base/*.t t/Business/DK/.svn/tmp/props/*.t t/Business/DK/.svn/tmp/text-base/*.t t/Business/DK/FI/*.t t/Business/DK/FI/.svn/*.t t/Business/DK/FI/.svn/prop-base/*.t t/Business/DK/FI/.svn/props/*.t t/Business/DK/FI/.svn/text-base/*.t t/Business/DK/FI/.svn/tmp/*.t t/Business/DK/FI/.svn/tmp/prop-base/*.t t/Business/DK/FI/.svn/tmp/props/*.t t/Business/DK/FI/.svn/tmp/text-base/*.t t/Class/*.t t/Class/.svn/*.t t/Class/.svn/prop-base/*.t t/Class/.svn/props/*.t t/Class/.svn/text-base/*.t t/Class/.svn/tmp/*.t t/Class/.svn/tmp/prop-base/*.t t/Class/.svn/tmp/props/*.t t/Class/.svn/tmp/text-base/*.t t/Class/Business/*.t t/Class/Business/.svn/*.t t/Class/Business/.svn/prop-base/*.t t/Class/Business/.svn/props/*.t t/Class/Business/.svn/text-base/*.t t/Class/Business/.svn/tmp/*.t t/Class/Business/.svn/tmp/prop-base/*.t t/Class/Business/.svn/tmp/props/*.t t/Class/Business/.svn/tmp/text-base/*.t t/Class/Business/DK/*.t t/Class/Business/DK/.svn/*.t t/Class/Business/DK/.svn/prop-base/*.t t/Class/Business/DK/.svn/props/*.t t/Class/Business/DK/.svn/text-base/*.t t/Class/Business/DK/.svn/tmp/*.t t/Class/Business/DK/.svn/tmp/prop-base/*.t t/Class/Business/DK/.svn/tmp/props/*.t t/Class/Business/DK/.svn/tmp/text-base/*.t t/Class/Business/DK/FI/*.t t/Class/Business/DK/FI/.svn/*.t t/Class/Business/DK/FI/.svn/prop-base/*.t t/Class/Business/DK/FI/.svn/props/*.t t/Class/Business/DK/FI/.svn/text-base/*.t t/Class/Business/DK/FI/.svn/tmp/*.t t/Class/Business/DK/FI/.svn/tmp/prop-base/*.t t/Class/Business/DK/FI/.svn/tmp/props/*.t t/Class/Business/DK/FI/.svn/tmp/text-base/*.t'
            }
)
;


lib/Business/DK/FI.pm
package Business::DK::FI;

# $Id: FI.pm 8172 2013-07-31 16:45:55Z jonasbn $

use strict;
use warnings;
use vars qw($VERSION @EXPORT_OK);
use Params::Validate qw(validate_pos SCALAR ARRAYREF);
use Readonly;
use base qw(Exporter);
use English qw( -no_match_vars );

$VERSION   = '0.06';
@EXPORT_OK = qw(validate validateFI generate);

use constant MODULUS_OPERAND => 10;
use constant THRESHOLD       => 10;
use constant DEDUCTION       => 9;
use constant INVALID         => 0;
use constant VALID           => 1;

Readonly::Array my @CONTROLCIFERS  => qw(1 2 1 2 1 2 1 2 1 2 1 2 1 2);
Readonly::Scalar my $CONTROL_LENGTH => scalar @CONTROLCIFERS;

## no critic (NamingConventions::Capitalization)

sub validateFI {
    return validate(shift);
}

sub validate {
    my ($fi_number) = @ARG;

    validate_pos( @ARG, { type => SCALAR, regex => qr/^\d{15}$/xsm } );

    my ($last_digit);
    ( $fi_number, $last_digit )
        = $fi_number =~ m/^(\d{$CONTROL_LENGTH})(\d{1})$/xsm;

    my $sum = _calculate_sum( $fi_number, \@CONTROLCIFERS );
    my $checksum = _calculate_checksum($sum);

    if ( $checksum == $last_digit ) {
        return VALID;
    } else {
        return INVALID;
    }
}

sub _calculate_checksum {
    my ($sum) = @ARG;

    validate_pos( @ARG, { type => SCALAR, regex => qr/^\d+$/xsm }, );

    return ( THRESHOLD - ( $sum % MODULUS_OPERAND ) );
}

sub _calculate_sum {
    my ( $number, $CONTROLCIFERS ) = @ARG;

    validate_pos(
        @ARG,
        { type => SCALAR, regex => qr/^\d+$/xsm },
        { type => ARRAYREF },
    );

    my $sum = 0;
    my @numbers = split //smx, $number;

## no critic (ControlStructures::ProhibitCStyleForLoops)
    for ( my $i = 0; $i < scalar @numbers; $i++ ) {
        my $tmp_sum = $numbers[$i] * $CONTROLCIFERS->[$i];

        if ( $tmp_sum >= THRESHOLD ) {
            $sum += ( $tmp_sum - DEDUCTION );
        } else {
            $sum += $tmp_sum;
        }
    }
    return $sum;
}

sub generate {
    my ($number) = @ARG;

    #number has to be a positive number between 1 and 99999999999999
    validate_pos(
        @ARG,
        {   type      => SCALAR,
            regex     => qr/^\d+$/,
            callbacks => {
                'higher than 0' => sub { shift() >= 1 },
                'lower than 99999999999999' =>
                    sub { shift() <= 99999999999999 },
            },
        },
    );

    #padding with zeroes up to our maximum length
    my $pattern = '%0' . $CONTROL_LENGTH . 's';
    my $reformatted_number = sprintf $pattern, $number;

    #this call takes care of the check of the product of the above statement
    my $sum = _calculate_sum( $reformatted_number, \@CONTROLCIFERS );
    my $checksum = _calculate_checksum($sum);

    my $finalized_number = $reformatted_number . $checksum;

    return $finalized_number;
}

1;

__END__

=head1 NAME

Business::DK::FI - validation of Danish FI numbers

=head1 VERSION

The documentation describes version 0.02

=head1 SYNOPSIS

    use Business::DK::FI qw(validate validateFI generate);
    
    if (validate('026840149965328')) {
        print "026840149965328 is valid\n";
    }
    
    
    my $fi_number = generate(1);
    
    if ($fi_number eq '000000000000018') {
        print "we have a FI number\n";
    }
    

=head1 DESCRIPTION

FI numbers are numbers used on GIRO payment forms. These can be used to do
online payments in banks or at in physical banks or post offices.

The module currently only supports FI numbers in the following series:

=over

=item * 71

=item * 75

=back

=head1 SUBROUTINES AND METHODS

=head2 validate

Takes a single argument. 15 digit FI number. Returns true (1) or false (0)
indicating whether the provided parameter adheres to specification.

=head2 validateFI

Less intrusive exported variation of L</validate>. It is actually L</validate>
which is wrapping L</validateFI>.

=head2 generate

Simple FI generation method. Takes an arbitrary number adhering to the following requirements: 

=over

=item * length between 1 and 14

=item * value between 1 and 99999999999999

=back

Returns a valid FI number.

=head2 PRIVATE SUBROUTINES AND METHODS

=head2 _calculate_checksum

This method calculates a checksum, it takes a single number as parameter and returns the calculated checksum.

=head2 _calculate_sum

This method calculates a sum it takes a number and a reference to an array of control cifers. It calculates a single sum based on the number and the control cifer and returns this.

=head1 DIAGNOSTICS

All methods B<die> if their API is not respected. Method calls can with success be wrapped in L<Try::Tiny> or C<eval> blocks.

=head1 CONFIGURATION AND ENVIRONMENT

The module requires no special configuration or environment.

=head1 DEPENDENCIES

=over

=item * L<Params::Validate>

=item * L<Readonly>

=item * L<Exporter>

=item * L<English>

=back

=head1 BUGS AND LIMITATIONS

This module has no known bugs or limitations.

=head1 BUG REPORTING

Please report issues via CPAN RT:

=over

=item * L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-DK-FI>

=back

or by sending mail to

=over

=item * C<< <bug-Business-DK-FI@rt.cpan.org> >>

=back

=head1 TEST AND QUALITY

=head2 TEST COVERAGE

    ---------------------------- ------ ------ ------ ------ ------ ------ ------
    File                           stmt   bran   cond    sub    pod   time  total
    ---------------------------- ------ ------ ------ ------ ------ ------ ------
    blib/lib/Business/DK/FI.pm    100.0  100.0    n/a  100.0  100.0   34.8  100.0
    ...b/Class/Business/DK/FI.pm   97.6   83.3    n/a  100.0  100.0   65.2   96.9
    Total                          99.1   90.0    n/a  100.0  100.0  100.0   98.7
    ---------------------------- ------ ------ ------ ------ ------ ------ ------

=head1 QUALITY AND CODING STANDARD

The code passes L<Perl::Critic> tests at severity 1 (I<brutal>) with a set of policies disabled. please see F<t/perlcriticrc> and the list below:

=over

=item * L<Perl::Critic::Policy::ValuesAndExpressions::ProhibitConstantPragma>, please
see: L<http://logiclab.jira.com/wiki/display/OPEN/Perl-Critic-Policy-ValuesAndExpressions-ProhibitConstantPragma>

=item * L<Perl::Critic::Policy::Documentation::RequirePodLinksIncludeText>, this
is listed in the L</TODO> it requires numerous changes to the distribution POD.

=item * L<Perl::Critic::Policy::NamingConventions::Capitalization>, L</validateFI> is exported both as L</validateFI> and L</validate> and B<FI> is our used acronym
so we try to stick to this

=item * L<Perl::Critic::Policy::ControlStructures::ProhibitCStyleForLoops>, this
is for the main algorithm, it was easier to do with a B<C-style for loop>

=item * L<Perl::Critic::Policy::Subroutines::RequireArgUnpacking>, this is due to
the way: L<Params::Validate> is used

=item * L<Perl::Critic::Policy::ValuesAndExpressions::ProhibitMagicNumbers>, we
use long control numbers

=item * L<Perl::Critic:.:Policy::Variables::ProhibitPunctuationVars>

=back

=head1 TODO

Please see the distribution F<TODO> file also and the distribution road map at:
    L<http://logiclab.jira.com/browse/BDKFI#selectedTab=com.atlassian.jira.plugin.system.project%3Aroadmap-panel>

=head1 SEE ALSO

=over

=item * http://www.pbs.dk/

=item * L<Try::Tiny>

=item * L<Business::DK::CVR>

=item * L<Business::DK::CPR>

=item * L<Business::DK::PO>

=item * L<Business::DK::Postalcode>

=item * L<Business::DK::Phonenumber>

=back

=head1 AUTHOR

Jonas B. Nielsen, (jonasbn) - C<< <jonasbn@cpan.org> >>

=head1 COPYRIGHT

Business-DK-FI and related is (C) by Jonas B. Nielsen, (jonasbn) 2009-2011

=head1 LICENSE

Business-DK-FI and related is released under the artistic license

The distribution is licensed under the Artistic License, as specified
by the Artistic file in the standard perl distribution
(http://dev.perl.org/licenses/artistic.html).

=cut
